### PR TITLE
Rather simpler

### PR DESCRIPTION
**Pull request by treeowl**
Don't use a list at all. Just follow the structure from the start.

### CODE REVIEW

**Comment by treeowl** [comment]
Ooh... Maybe I can use `StateT`....

**Comment by treeowl** [comment]
Not sure if `StateT` is better or worse, but I'm leaning toward slightly worse. Let me know if you have a preference.

**Comment by treeowl** [comment]
Yeah... `StateT` is overkill in the wrong direction. Forget that.

File: src/Language/Haskell/TH/Lift/Generics.hs [hunk 1]
```diff
@@ -251,27 +250,32 @@ instance (GLiftDatatype f, GLiftDatatype g) => GLiftDatatype (f :+: g) where
...
+    -- | @gliftArgs e f@ applies @f@ to the zero or more arguments represented
+    -- by @e@.
+    gliftArgs :: Quote m => f a -> Exp -> m Exp
 
 instance GLiftArgs U1 where
-    gliftArgs U1 = []
+    -- This pattern match must be strict, because
+    -- lift undefined really shouldn't just happen
+    -- to work for unit types.
+    gliftArgs U1 = return
```
Comment (treeowl) [inline src/Language/Haskell/TH/Lift/Generics.hs:276]: Should we use something like

```haskell
gliftArgs u = \x -> return (u `seq` x)
```

to match the behavior of the `V1` instance? Or (more to my taste) should we make the `V1` instance stricter to match `U1`? I don't see any *obvious* benefit to delaying the match.

Comment (RyanGlScott) [reply src/Language/Haskell/TH/Lift/Generics.hs:276]: I'm not quite sure I understand the question, as there's no `GLiftArgs` instance for `V1`. Moreover, I'm unclear on what the difference between `gliftArgs U1 = return` and `gliftArgs u = \x -> return (u `seq` x)` is. I think I'd need to see concrete examples that demonstrate how they differ.

**Review by RyanGlScott — COMMENTED** [review]
Thanks!

**Comment by treeowl** [comment]
The current behavior:

```haskell
genericLift (undefined @Void) = pure undefined
genericLift (undefined @()) = pure undefined
```

The behavior in this PR:

```haskell
genericLift (undefined @Void) = pure undefined
genericLift (undefined @()) = undefined
```

Not very satisfying, IMO. Unless you have a specific reason to want the current behavior, I would recommend the following:

```haskell
genericLift (undefined @Void) = undefined
genericLift (undefined @()) = undefined
```

Make sense now?

**Comment by treeowl** [comment]
I've added another commit with the proposed `V1` change.

**Comment by treeowl** [comment]
Sorry, I got that a bit wrong, didn't I? *sigh*. The current behavior is actually the same as the PR, thanks to the `sequence` bit. I don't like the `V1` instance.

**Comment by treeowl** [comment]
Latest commit changes `V1`, though I doubt anyone will notice.

**Comment by RyanGlScott** [comment]
Ah, I suppose you were referring to the `GLiftDatatype` instance for `V1`, not `GLiftArgs`. (My apologies, that threw me for a loop.)

As far as how lazy we should be, my inclination is to look to `DeriveLift` as inspiration. Based on the `-ddump-simpl` output for this program:

```hs
{-# LANGUAGE DeriveLift #-}
module Foo where

import Language.Haskell.TH.Syntax (Lift)

data V1 deriving Lift
data U1 = U1 deriving Lift
```
```
==================== Derived instances ====================
Derived class instances:
  instance Language.Haskell.TH.Syntax.Lift Foo.U1 where
    Language.Haskell.TH.Syntax.lift Foo.U1 = [| Foo.U1 |]
    Language.Haskell.TH.Syntax.liftTyped Foo.U1 = [|| Foo.U1 ||]
  
  instance Language.Haskell.TH.Syntax.Lift Foo.V1 where
    Language.Haskell.TH.Syntax.lift z_a1Ym
      = GHC.Base.pure (case z_a1Ym of)
    Language.Haskell.TH.Syntax.liftTyped z_a1Yn
      = Language.Haskell.TH.Syntax.unsafeCodeCoerce
          (GHC.Base.pure (case z_a1Yn of))
```

If I'm looking at this correctly, this means that `lift (undefined :: U1) = undefined` and `lift (undefined :: V1) = pure undefined`. The latter behavior, I believe, resulted from a conscious design choice in ghc-proposals/ghc-proposals#63 to make instances as maximally lazy as possible. In this regard, the behavior for `U1` is inconsistent, but then again, GHC's `deriving` behavior for unit data constructors wasn't put through the proposal process in the same way. ([This section](https://www.haskell.org/onlinereport/haskell2010/haskellch11.html#x18-18300011.1) of the 2010 Haskell Report states that derived instances of `Eq` and `Ord` are always strict in both their arguments, so perhaps other derived instances took after that.)

At the end of the day, I don't feel too strongly about this. As long as we document the behavior somewhere, I'm fine with whatever design choice you settle on.

**Comment by treeowl** [comment]
I've never really been comfortable with the maximal laziness decision there, but I'm not going to fight it in general at the moment. However, it's particularly hard to see any value in this context—would someone ever perform the monadic action `lift` returns and then throw away the result?

**Review by RyanGlScott — APPROVED** [review]

**Comment by treeowl** [comment]
Yay! Thanks.

**Pull request merged**